# Practice Loading and Exploring Datasets

This assignment is purposely open-ended. You will be asked to load datasets from the [UC-Irvine Machine Learning Repository](https://archive.ics.uci.edu/ml/index.php). 

Even though you maybe using different datasets than your fellow classmates, try and be supportive and assist each other in the challenges that you are facing. You will only deepend your understanding of these topics as you work to assist one another. Many popular UCI datasets face similar data loading challenges.

Remember that the UCI datasets do not necessarily have a file type of `.csv` so it's important that you learn as much as you can about the dataset before you try and load it. See if you can look at the raw text of the file either locally or using the `!curl` shell command -or in some other way before you try and read it in as a dataframe. This will help you catch what would otherwise be unforseen problems.

Feel free to embellish this notebook with additional markdown cells,code cells, comments, graphs, etc. Whatever you think helps adequately address the questions.


Edit: Kilroy was here

## 1) Load a dataset from UCI (via its URL)

Please navigate to the home page and choose a dataset (other than the Adult dataset) from the "Most Popular" section on the right-hand side of the home page. Load the dataset via its URL and check the following (show your work):

- Are the headers showing up properly?
- Look at the first 5 and the last 5 rows, do they seem to be in order?
- Does the dataset have the correct number of rows and columns as described in the UCI page? 
 - Remember, that UCI does not count the y variable (column of values that we might want to predict via a machine learning model) as an "attribute" but rather as a "class attribute" so you may end up seeing a number of columns that is one greater than the number listed on the UCI website.
- Does UCI list this dataset as having missing values? Check for missing values and see if your analysis corroborates what UCI reports?
- if `NaN` values or other missing value indicators are not being detected by `df.isnull().sum()` find a way to replace whatever is indicating the missing values with `np.NaN`.
- Use the .describe() function in order to see the summary statistics of both the numeric and non-numeric columns. 

In [0]:
#Answers to 1)

abaloneurl = 'https://archive.ics.uci.edu/ml/machine-learning-databases/abalone/abalone.data'

#Check the file with curl
#!curl https://archive.ics.uci.edu/ml/machine-learning-databases/abalone/abalone.data


In [0]:
#Load the file via URL
import pandas as pd
abalone = pd.read_csv(abaloneurl, header=None)

### Are the headers showing up correctly?

In [114]:
abalone.head()

0      1      2      3       4       5       6      7   8
0  M  0.455  0.365  0.095  0.5140  0.2245  0.1010  0.150  15
1  M  0.350  0.265  0.090  0.2255  0.0995  0.0485  0.070   7
2  F  0.530  0.420  0.135  0.6770  0.2565  0.1415  0.210   9
3  M  0.440  0.365  0.125  0.5160  0.2155  0.1140  0.155  10
4  I  0.330  0.255  0.080  0.2050  0.0895  0.0395  0.055   7


**The headers are showing up correctly but not named.**

In [0]:
#Code to fix headers in abalone dataframe
abalone_headers = ['sex', 'length', 'diameter', 'height', 'whole_weight', 'shucked_weight', 'viscera_weight', 'shell weight', 'rings']

abalone = pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/abalone/abalone.data', header = None, names= abalone_headers)

### Look at the first 5 and the last 5 rows, do they seem to be in order?

In [108]:
#first five rows
abalone.head(5)


sex  length  diameter  ...  viscera_weight  shell weight  rings
0   M   0.455     0.365  ...          0.1010         0.150     15
1   M   0.350     0.265  ...          0.0485         0.070      7
2   F   0.530     0.420  ...          0.1415         0.210      9
3   M   0.440     0.365  ...          0.1140         0.155     10
4   I   0.330     0.255  ...          0.0395         0.055      7

[5 rows x 9 columns]

In [0]:
#Last five rows
abalone.tail(5)

sex  length  diameter  ...  viscera_weight  shell weight  rings
4172   F   0.565     0.450  ...          0.2390        0.2490     11
4173   M   0.590     0.440  ...          0.2145        0.2605     10
4174   M   0.600     0.475  ...          0.2875        0.3080      9
4175   F   0.625     0.485  ...          0.2610        0.2960     10
4176   M   0.710     0.555  ...          0.3765        0.4950     12

[5 rows x 9 columns]

**It doesn't look like they are in any sort of order... just a numbered list of individual abalone.**

### Does the dataset have the correct number of rows and columns as described in the UCI page?

In [0]:
abalone.shape

(4177, 9)

**According to the key, the file should have 4,177 rows an 8 columns. My shape file says 9 columns and 4177 rows. The numbers of rows match... The instructions note there may be 1 extra column and there is!**

### Does UCI list this dataset as having missing values? Check for missing values and see if your analysis corroborates what UCI reports?

*UCI lists this dataset as NOT having any missing values.*

In [0]:
abalone.isna().sum()

sex               0
length            0
diameter          0
height            0
whole_weight      0
shucked_weight    0
viscera_weight    0
shell weight      0
rings             0
dtype: int64

**According to the `is.na().sum()' call on `abalone`... there are 0 missing values across all attributes. This matches the UCI database.**

### If NaN values or other missing value indicators are not being detected by df.isnull().sum() find a way to replace whatever is indicating the missing values with np.NaN.

**All of the missing values are being detected correctly.**

### Use the .describe() function in order to see the summary statistics of both the numeric and non-numeric columns.

In [0]:
#Summary statistics for numeric values
abalone.describe()

length     diameter  ...  shell weight        rings
count  4177.000000  4177.000000  ...   4177.000000  4177.000000
mean      0.523992     0.407881  ...      0.238831     9.933684
std       0.120093     0.099240  ...      0.139203     3.224169
min       0.075000     0.055000  ...      0.001500     1.000000
25%       0.450000     0.350000  ...      0.130000     8.000000
50%       0.545000     0.425000  ...      0.234000     9.000000
75%       0.615000     0.480000  ...      0.329000    11.000000
max       0.815000     0.650000  ...      1.005000    29.000000

[8 rows x 8 columns]

In [0]:
#Summary statistics for non-numerics
abalone.describe(exclude = 'number')

sex
count   4177
unique     3
top        M
freq    1528

## 2) Load a dataset from your local machine.
Choose a second dataset from the "Popular Datasets" listing on UCI, but this time download it to your local machine instead of reading it in via the URL. Upload the file to Google Colab using the files tab in the left-hand sidebar or by importing `files` from `google.colab` The following link will be a useful resource if you can't remember the syntax: <https://towardsdatascience.com/3-ways-to-load-csv-files-into-colab-7c14fcbdcb92>

- Answer all of the same bullet point questions from part 1 again on this new dataset. 


In [71]:
# Using Car Evaluation dataset
from google.colab import files
uploaded = files.upload()

Saving car.data to car (1).data


In [0]:
import io
cars_data = pd.read_csv(io.BytesIO(uploaded['car.data']), header = None)

### Are the headers showing up correctly?

In [133]:
cars_data.head()

buying  maint doors persons lug_boot safety comfort
0  vhigh  vhigh     2       2    small    low   unacc
1  vhigh  vhigh     2       2    small    med   unacc
2  vhigh  vhigh     2       2    small   high   unacc
3  vhigh  vhigh     2       2      med    low   unacc
4  vhigh  vhigh     2       2      med    med   unacc

The headers are showing up but not named.


In [0]:
car_names = ['buying', 'maint', 'doors', 'persons', 'lug_boot', 'safety', 'comfort']
cars_data = pd.read_csv(io.BytesIO(uploaded['car.data']), header=None, names= car_names)

**Names should be fixed (note: there seem to only be 6 names in the dictionary but 7 attributes?**

### Look at the first 5 and the last 5 rows, do they seem to be in order?

In [134]:
cars_data.head(5)

buying  maint doors persons lug_boot safety comfort
0  vhigh  vhigh     2       2    small    low   unacc
1  vhigh  vhigh     2       2    small    med   unacc
2  vhigh  vhigh     2       2    small   high   unacc
3  vhigh  vhigh     2       2      med    low   unacc
4  vhigh  vhigh     2       2      med    med   unacc

In [118]:
cars_data.tail(5)

0    1      2     3    4     5      6
1723  low  low  5more  more  med   med   good
1724  low  low  5more  more  med  high  vgood
1725  low  low  5more  more  big   low  unacc
1726  low  low  5more  more  big   med   good
1727  low  low  5more  more  big  high  vgood

**It looks like maybe they are sorted by the first attribute (price)**

### Does the dataset have the correct number of rows and columns as described in the UCI page?

In [135]:
cars_data.shape

(1728, 7)

**UCI documentation says the dataset should have 6 attributes and 1728 rows. It seems to match given the extra column mentioned in the instructions.**

### Does UCI list this dataset as having missing values? Check for missing values and see if your analysis corroborates what UCI reports?

In [136]:
cars_data.isna().sum()

buying      0
maint       0
doors       0
persons     0
lug_boot    0
safety      0
comfort     0
dtype: int64

**The UCI documentation says there are no missing values and the `isna().sum()` call shows no missing values in any column.**

### If NaN values or other missing value indicators are not being detected by df.isnull().sum() find a way to replace whatever is indicating the missing values with np.NaN.

**All of the missing values are accounted for.**

### Use the .describe() function in order to see the summary statistics of both the numeric and non-numeric columns.

In [140]:
cars_data.describe()

buying maint doors persons lug_boot safety comfort
count    1728  1728  1728    1728     1728   1728    1728
unique      4     4     4       3        3      3       4
top       low   low     4       4    small    low   unacc
freq      432   432   432     576      576    576    1210

In [138]:
cars_data.describe(exclude='number')

buying maint doors persons lug_boot safety comfort
count    1728  1728  1728    1728     1728   1728    1728
unique      4     4     4       3        3      3       4
top       low   low     4       4    small    low   unacc
freq      432   432   432     576      576    576    1210

**I guess it's nothing but categorical variables!**

## 3) Make Crosstabs of the Categorical Variables

Take whichever of the above datasets has more categorical variables and use crosstabs to tabulate the different instances of the categorical variables.


In [0]:
# Your Code Here

## 4) Explore the distributions of the variables of the dataset using:
- Histograms
- Scatterplots
- Density Plots

In [0]:
# Your Code Here

## 5) Create at least one visualization from a crosstab:

Remember that a crosstab is just a dataframe and can be manipulated in the same way by row index, column, index, or column/row/cell position.


In [0]:
# Your Code Here

## Stretch Goals 

The following additional study tasks are optional, they are intended to give you an opportunity to stretch yourself beyond the main requirements of the assignment. You can pick and choose from the below, you do not need to complete them in any particular order.

### - Practice Exploring other Datasets

### -  Try using the Seaborn plotting library's "Pairplot" functionality in order to explore all of the possible histograms and scatterplots of your dataset all at once:

[Seaborn Pairplot](https://seaborn.pydata.org/generated/seaborn.pairplot.html)

### - Turn some of the continuous variables into categorical variables by binning the values using:
- [pd.cut()](https://pandas.pydata.org/pandas-docs/version/0.23.4/generated/pandas.cut.html)
- [pd.qcut()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.qcut.html)
- <https://stackoverflow.com/questions/30211923/what-is-the-difference-between-pandas-qcut-and-pandas-cut>

And then use crosstabs to compare/visualize these binned variables against the other variables.


### - Other types and sources of data
Not all data comes in a nice single file - for example, image classification involves handling lots of image files. You still will probably want labels for them, so you may have tabular data in addition to the image blobs - and the images may be reduced in resolution and even fit in a regular csv as a bunch of numbers.

If you're interested in natural language processing and analyzing text, that is another example where, while it can be put in a csv, you may end up loading much larger raw data and generating features that can then be thought of in a more standard tabular fashion.

Overall you will in the course of learning data science deal with loading data in a variety of ways. Another common way to get data is from a database - most modern applications are backed by one or more databases, which you can query to get data to analyze. We'll cover this more in our data engineering unit.

How does data get in the database? Most applications generate logs - text files with lots and lots of records of each use of the application. Databases are often populated based on these files, but in some situations you may directly analyze log files. The usual way to do this is with command line (Unix) tools - command lines are intimidating, so don't expect to learn them all at once, but depending on your interests it can be useful to practice.

One last major source of data is APIs: https://github.com/toddmotto/public-apis

API stands for Application Programming Interface, and while originally meant e.g. the way an application interfaced with the GUI or other aspects of an operating system, now it largely refers to online services that let you query and retrieve data. You can essentially think of most of them as "somebody else's database" - you have (usually limited) access.

*Stretch goal* - research one of the above extended forms of data/data loading. See if you can get a basic example working in a notebook. Image, text, or (public) APIs are probably more tractable - databases are interesting, but there aren't many publicly accessible and they require a great deal of setup.